File: mobile/src/main/java/treehou/se/habit/ui/widgets/factories/colorpicker/HoldListener.java
package treehou.se.habit.ui.widgets.factories.colorpicker;


import android.view.MotionEvent;
import android.view.View;

import org.jetbrains.annotations.NotNull;

import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;

class HoldListener implements View.OnTouchListener {

    private static final int DEFAULT_TICK_TIME=200;

    private int tick=0;
    private Observable<Long> timer;
    private Disposable disposable = null;
    private OnHoldListener listener;
    private Consumer<Long> touchSubject;

    public HoldListener(@NotNull OnHoldListener listener) {
        this(listener, DEFAULT_TICK_TIME);
    }

    public HoldListener(@NotNull final OnHoldListener listener, int tickTime) {
        this.listener = listener;

        timer = Observable.interval(tickTime, TimeUnit.MILLISECONDS);

        touchSubject = time -> updateTick();
        timer = Observable.interval(tickTime, TimeUnit.MILLISECONDS).doOnDispose(() -> listener.onRelease(tick));
    }

    @Override
    public boolean onTouch(View v, MotionEvent event) {

        if(event.getAction() == MotionEvent.ACTION_DOWN) {
            disposable = timer.subscribe(touchSubject);
        } else if (event.getAction() == MotionEvent.ACTION_UP ||
                event.getAction() == MotionEvent.ACTION_CANCEL ||
                event.getAction() == MotionEvent.ACTION_OUTSIDE) {

            if (!disposable.isDisposed()) {
                disposable.dispose();
            }
        }
        return true;
    }

    private void updateTick(){
        tick++;
        listener.onTick(tick);
    }

    interface OnHoldListener {
        void onTick(int tick);
        void onRelease(int tick);
    }
}

File: mobile/src/main/java/treehou/se/habit/ui/widget/HoldListener.kt
package treehou.se.habit.ui.widget


import android.annotation.SuppressLint
import android.view.MotionEvent
import android.view.View

import java.util.concurrent.TimeUnit

import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Consumer

class HoldListener @JvmOverloads constructor(private val tickListener: OnTickListener, private val releaseListener: OnReleaseListener, tickTime: Int = DEFAULT_TICK_TIME) : View.OnTouchListener {

    private var tick = 0
    private var timer: Observable<Long>? = null
    private var disposable: Disposable? = null
    private val touchSubject: Consumer<Long>

    init {

        timer = Observable.interval(tickTime.toLong(), TimeUnit.MILLISECONDS)
        touchSubject = Consumer { updateTick() }
        timer = Observable.interval(tickTime.toLong(), TimeUnit.MILLISECONDS).doOnDispose { releaseListener.onRelease(tick) }
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouch(v: View, event: MotionEvent): Boolean {

        if (event.action == MotionEvent.ACTION_DOWN) {
            disposable = timer!!.subscribe(touchSubject)
        } else if (event.action == MotionEvent.ACTION_UP ||
                event.action == MotionEvent.ACTION_CANCEL ||
                event.action == MotionEvent.ACTION_OUTSIDE) {

            if (!disposable!!.isDisposed) {
                disposable!!.dispose()
            }
        }
        return true
    }

    private fun updateTick() {
        tick++
        tickListener.onTick(tick)
    }

    interface OnTickListener {
        fun onTick(tick: Int)
    }

    interface OnReleaseListener {
        fun onRelease(tick: Int)
    }

    companion object {

        private val DEFAULT_TICK_TIME = 200
    }
}
